network/p2p/router.go

```diff
@@ -0,0 +1,232 @@
+
+	_ common.AppHandler    = (*Router)(nil)
+	_ validators.Connector = (*Router)(nil)
+)
+
+// Router routes incoming application messages to the corresponding registered
+// app handler. App messages must be made using the registered handler's
+// corresponding Client.
+type Router struct {
+	log    logging.Logger
+	sender common.AppSender
+
+	lock                         sync.RWMutex
+	handlers                     map[uint64]*responder
+	pendingAppRequests           map[uint32]AppResponseCallback
+	pendingCrossChainAppRequests map[uint32]CrossChainAppResponseCallback
+	requestID                    uint32
+	peers                        set.SampleableSet[ids.NodeID]
+}
+
+// NewRouter returns a new instance of Router
+func NewRouter(log logging.Logger, sender common.AppSender) *Router {
+	return &Router{
+		log:                          log,
+		sender:                       sender,
+		handlers:                     make(map[uint64]*responder),
+		pendingAppRequests:           make(map[uint32]AppResponseCallback),
+		pendingCrossChainAppRequests: make(map[uint32]CrossChainAppResponseCallback),
+	}
+}
+
+func (r *Router) Connected(_ context.Context, nodeID ids.NodeID, _ *version.Application) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	r.peers.Add(nodeID)
+	return nil
+}
+
+func (r *Router) Disconnected(_ context.Context, nodeID ids.NodeID) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	r.peers.Remove(nodeID)
+	return nil
+}
+
+// RegisterAppProtocol reserves an identifier for an application protocol and
+// returns a Client that can be used to send messages for the corresponding
+// protocol.
+func (r *Router) RegisterAppProtocol(handlerID uint64, handler Handler) (*Client, error) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if _, ok := r.handlers[handlerID]; ok {
+		return nil, fmt.Errorf("failed to register handler id %d: %w", handlerID, ErrExistingAppProtocol)
+	}
+
+	r.handlers[handlerID] = &responder{
+		handlerID: handlerID,
+		handler:   handler,
+		log:       r.log,
+		sender:    r.sender,
+	}
+
+	return &Client{
+		handlerPrefix: binary.AppendUvarint(nil, handlerID),
+		sender:        r.sender,
+		router:        r,
+	}, nil
+}
+
+func (r *Router) AppRequest(ctx context.Context, nodeID ids.NodeID, requestID uint32, deadline time.Time, request []byte) error {
+	parsedMsg, handler, ok := r.parse(request)
+	if !ok {
+		r.log.Debug("failed to process message",
+			zap.Stringer("messageOp", message.AppRequestOp),
+			zap.Stringer("nodeID", nodeID),
+			zap.Uint32("requestID", requestID),
+			zap.Time("deadline", deadline),
+			zap.Binary("message", request),
+		)
+		return nil
+	}
+
+	return handler.AppRequest(ctx, nodeID, requestID, deadline, parsedMsg)
+}
+
+func (r *Router) AppRequestFailed(_ context.Context, nodeID ids.NodeID, requestID uint32) error {
+	callback, ok := r.clearAppRequest(requestID)
+	if !ok {
+		return ErrUnrequestedResponse
+	}
+
+	callback(nodeID, nil, ErrAppRequestFailed)
```

Comment:
it seems both for AppResponseCallback and CrossChainAppResponseCallback we could:
- pass context to `callback`
- return error from `callback`